GH issue #430: added links to kube docs on liveness checks

bfallonf opened the pull request:
@thoraxe @sosiouxme @nak3 I added the links to the Kube docs. I think it's more practical, as I'd just be retyping the Kube docs into our docs. i also cleaned up the doc a little.

If you disagree let me know. I can see the argument for having the examples in there.

Thanks.

bfallonf commented:
Might need to add more info to the Deployments section and links for the second option, as per the above closed issue.

adellape commented on dev_guide/application_health.adoc:
```diff
@@ -11,31 +11,37 @@ toc::[]
+In software systems, components can become unhealthy due to transient issues
+(such as temporary connectivity loss), configuration errors, or problems with
+external dependencies. OpenShift applications have a number of options to detect
+and handle unhealthy containers
```
Punctuation.

adellape commented on dev_guide/application_health.adoc:
```diff
@@ -11,31 +11,37 @@ toc::[]
+external dependencies. OpenShift applications have a number of options to detect
+and handle unhealthy containers
 
 == Liveness checks
```
Title case

adellape commented on dev_guide/application_health.adoc:
```diff
@@ -11,31 +11,37 @@ toc::[]
+methods to establish container liveness and policies that determine how 'dead'
+containers are handled.
+
+There are link:http://kubernetes.io/v1.0/docs/user-guide/liveness/[two types] of
```
Don't need `link:` here.

adellape commented on dev_guide/application_health.adoc:
```diff
@@ -11,31 +11,37 @@ toc::[]
 
-// TODO: example of liveness check
+For more information on health checks, see the
+link:http://kubernetes.io/v1.0/docs/user-guide/walkthrough/k8s201.html#health-checking[Kubernetes
```
Same re: `link:`.

adellape commented on dev_guide/application_health.adoc:
```diff
@@ -11,31 +11,37 @@ toc::[]
+link:http://kubernetes.io/v1.0/docs/user-guide/walkthrough/k8s201.html#health-checking[Kubernetes
+documentation].
 
 == High-level application health checks
```
Title case

adellape commented on dev_guide/application_health.adoc:
```diff
@@ -11,31 +11,37 @@ toc::[]
-Each component in a distributed system has a notion of liveness scoped to that component.  Liveness
-is the quality of being correctly configured and operational on an individual level.  To express
-this, OpenShift offers a number of methods to establish container liveness and policies that
-determine how 'dead' containers are handled.
```
I think the original ordering of these sentences worked better.

smarterclayton commented:
We need to add the readiness check, and talk about how they interact.  Readiness is described in the deployment example doc.

thoraxe commented:
The Kube docs are a bit of a maze. I don't mind linking out for deeper info, but I think it behooves us to have our own examples and docs "locally".

bfallonf commented:
@thoraxe @smarterclayton OK. I've added in examples for the liveness and readiness probes, as well as some rewording from info from the Kubernetes docs.

@smarterclayton You mention in issue #777 that we should add something to the deployment docs. Do you mean something in deployment strategies on how the strategies depend on containers passing readiness checks? Something similar to the note box I've added here?

Let me know if there's anything else.

smarterclayton commented:
We should add the information from
https://github.com/openshift/origin/blob/master/examples/deployment/README.md

On Thu, Oct 15, 2015 at 12:15 AM, Brice Fallon-Freeman <
notifications@github.com> wrote:

> @thoraxe <https://github.com/thoraxe> @smarterclayton
> <https://github.com/smarterclayton> OK. I've added in examples for the
> liveness and readiness probes, as well as some rewording from info from the
> Kubernetes docs.
>
> @smarterclayton <https://github.com/smarterclayton> You mention in issue
> #777 <https://github.com/openshift/openshift-docs/issues/777> that we
> should add something to the deployment docs. Do you mean something in
> deployment strategies on how the strategies depend on containers passing
> readiness checks? Something similar to the note box I've added here?
>
> Let me know if there's anything else.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/openshift-docs/pull/1070#issuecomment-148276438>
> .
>

bfallonf commented:
@smarterclayton I've added some of that info into both the heath checks file and the deployments file. Some of it seemed to be rolling strategy-specific, but some generic, so I filled in both parts.

Would you say this addresses your initial issue? ( #777 )

adellape commented on dev_guide/application_health.adoc:
```diff
@@ -11,31 +11,93 @@ toc::[]
+running container. Currently, two types of probes exist, each serving a
+different purpose. Both probes can be configured in
+http://kubernetes.io/v1.0/docs/user-guide/liveness/[two ways]: HTTP checks and
+container execution checks.
```
This states there are two ways (HTTP checks, container exec), but the k8s doc link below and https://github.com/openshift/origin/blob/master/examples/deployment/README.md#readiness-checks says there are three (includes TCP connection checks).

adellape commented:
@bfallonf I think there's more info we can glean from the "Readiness Checks" section here: https://github.com/openshift/origin/blob/master/examples/deployment/README.md#readiness-checks

bfallonf commented:
@adellape Ugh I was looking at a different page to get the info. I see what you mean about the third option.

I'll see if I can talk to you tomorrow morning (my time) to see what you mean about the readiness doc. I'm not convinced there's anything more to add there...

adellape commented on dev_guide/application_health.adoc:
```diff
@@ -11,31 +11,125 @@ toc::[]
+different purpose:
 
-// TODO: example of liveness check
+[cols="2", width="80%"]
```
In local build, this table comes out lookin kinda weird to me with these options, mainly the `width="80%"`. Recommend playing with it some, maybe:

`[cols="1,4"]`

adellape commented on dev_guide/deployments.adoc:
```diff
@@ -286,12 +286,16 @@ deployment is created using the new tag value for the `helloworld` container.
 scenarios.
 
+A deployment strategy depends on
+link:../dev_guide/application_health.html#readiness-probe[readiness checks] to
```
`#readiness-probe` doesn't go anywhere.

adellape commented on dev_guide/deployments.adoc:
```diff
@@ -299,10 +303,15 @@ no strategy is specified on a deployment configuration.
 process.
 
+The rolling deployment strategy waits for pods to pass their
+link:../dev_guide/application_health.html#readiness-probe[readiness check]
```
Same re: `#readiness-probe`.

bfallonf replied:
Both changed to go to the topic in general and not specific part.

adellape commented on dev_guide/deployments.adoc:
```diff
@@ -299,10 +303,15 @@ no strategy is specified on a deployment configuration.
 
+The rolling deployment strategy waits for pods to pass their
+link:../dev_guide/application_health.html#readiness-probe[readiness check]
+before scaling down old components, and does not allow pods that don't pass
```
s/don't/do not/

adellape commented on dev_guide/application_health.adoc:
```diff
@@ -11,31 +11,125 @@ toc::[]
+----
+...
+livenessProbe:
+  tcpsocket
```
Should be camelcase with a colon:

`tcpSocket:`

https://docs.openshift.org/latest/rest_api/kubernetes_v1.html#v1-probe
https://docs.openshift.org/latest/rest_api/kubernetes_v1.html#v1-tcpsocketaction

bfallonf replied:
Yeah those are the docs I used to write it up. Done. Also, camel case is a weird term. :camel:

adellape commented on dev_guide/application_health.adoc:
```diff
@@ -11,31 +11,125 @@ toc::[]
+----
+====
+
+A TCP socket check is ideal for applications that don't start listening until
```
s/don't/do not/

adellape commented:
@bfallonf Some final nits inline. After those (particularly the fix to the TCP socket example), LGTM and would be ready for rebase/squash/merge.

bfallonf commented:
@adellape Ok. Should be good now. I did the above. And I changed the "viz" bit too.

sosiouxme commented on dev_guide/application_health.adoc:
```diff
@@ -11,33 +11,126 @@ toc::[]
+
+OpenShift users can specify high-level health checks and control the health
+check methodology by specifying a script to run, and the action to be taken when
+a health check has a negative result by creating an event.
```
After the nice examples in the previous section this section seems maddeningly short on details. Are high-level checks something OpenShift facilitates? Does it create Event API objects? How do you specify the check and the response? Or is this basically just "you have an API, you can script stuff yourself, w00t"?

bfallonf replied:
The answer to all of these questions is... I don't know. Because this issue was about the health checks and using the info from the Kubernetes doc to fix that up, I'll make another about this section and ask for examples, etc.

sosiouxme commented on dev_guide/deployments.adoc:
```diff
@@ -286,11 +286,14 @@ deployment is created using the new tag value for the `helloworld` container.
+A deployment strategy determines the deployment process, and is defined by the
+deployment configuration. Each application has different requirements for
+availability (and other considerations) during deployments. OpenShift provides
+out-of-the-box strategies to support a variety of deployment scenarios.
```
"out-of-the-box" seems redundant here.

bfallonf replied:
True. Deleted.

sosiouxme commented on dev_guide/deployments.adoc:
```diff
@@ -286,11 +286,14 @@ deployment is created using the new tag value for the `helloworld` container.
+availability (and other considerations) during deployments. OpenShift provides
+out-of-the-box strategies to support a variety of deployment scenarios.
+
+A deployment strategy adopts link:../dev_guide/application_health.html[readiness
```
"adopts" seems odd. How about just plain "uses"?

bfallonf replied:
Hmm I changed it to 'adopts' because it didn't feel right using 'uses' with 'used' at the end of the same sentence. Changed it back though. I can see both arguments.

bfallonf commented:
@sosiouxme Comments addressed. 

If nothing else, this should be ok to merge.

sosiouxme commented:
+1

adellape commented:
@bfallonf you're a :star:

bfallonf commented:
Thanks, all. FYI, here's the BZ:

https://bugzilla.redhat.com/show_bug.cgi?id=1274136
